# Mini-Project: JSON

## Import Modules and Files

In [30]:
import pandas as pd
import json
from pandas.io.json import json_normalize
import numpy as np

In [31]:
world_bank_df = pd.read_json('world_bank_projects.json')
world_bank_df.head()

_id  approvalfy board_approval_month  \
0  {'$oid': '52b213b38594d8a2be17c780'}        1999             November   
1  {'$oid': '52b213b38594d8a2be17c781'}        2015             November   
2  {'$oid': '52b213b38594d8a2be17c782'}        2014             November   
3  {'$oid': '52b213b38594d8a2be17c783'}        2014              October   
4  {'$oid': '52b213b38594d8a2be17c784'}        2014              October   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   
2  2013-11-01T00:00:00Z   MINISTRY OF FINANCE AND ECONOMIC DEVEL   
3  2013-10-31T00:00:00Z   MIN. OF PLANNING AND INT'L COOPERATION   
4  2013-10-31T00:00:00Z                      MINISTRY OF FINANCE   

            closingdate                              country_namecode  \
0  2018-07-07T00:00:00Z  Federal Democratic Republic of Ethiopia!$!ET   
1                   NaN                      Republic of Tunisia!$!TN   
2                   NaN                                   Tuvalu!$!TV   
3                   NaN                        Republic of Yemen!$!RY   
4  2019-04-30T00:00:00Z                       Kingdom of Lesotho!$!LS   

  countrycode                              countryname    countryshortname  \
0          ET  Federal Democratic Republic of Ethiopia            Ethiopia   
1          TN                      Republic of Tunisia             Tunisia   
2          TV                                   Tuvalu              Tuvalu   
3          RY                        Republic of Yemen  Yemen, Republic of   
4          LS                       Kingdom of Lesotho             Lesotho   

                         ...                           sectorcode source  \
0                        ...                          ET,BS,ES,EP   IBRD   
1                        ...                                BZ,BS   IBRD   
2                        ...                                   TI   IBRD   
3                        ...                                   JB   IBRD   
4                        ...                             FH,YW,YZ   IBRD   

   status  supplementprojectflg  \
0  Active                     N   
1  Active                     N   
2  Active                     Y   
3  Active                     N   
4  Active                     N   

                                              theme1  \
0      {'Percent': 100, 'Name': 'Education for all'}   
1  {'Percent': 30, 'Name': 'Other economic manage...   
2    {'Percent': 46, 'Name': 'Regional integration'}   
3  {'Percent': 50, 'Name': 'Participation and civ...   
4  {'Percent': 30, 'Name': 'Export development an...   

                                      theme_namecode    themecode   totalamt  \
0      [{'code': '65', 'name': 'Education for all'}]           65  130000000   
1  [{'code': '24', 'name': 'Other economic manage...        54,24          0   
2  [{'code': '47', 'name': 'Regional integration'...  52,81,25,47    6060000   
3  [{'code': '57', 'name': 'Participation and civ...        59,57          0   
4  [{'code': '45', 'name': 'Export development an...        41,45   13100000   

  totalcommamt                                                url  
0    130000000  http://www.worldbank.org/projects/P129828/ethi...  
1      4700000  http://www.worldbank.org/projects/P144674?lang=en  
2      6060000  http://www.worldbank.org/projects/P145310?lang=en  
3      1500000  http://www.worldbank.org/projects/P144665?lang=en  
4     13100000  http://www.worldbank.org/projects/P144933/seco...  

[5 rows x 50 columns]

## Problem 1: Fin Top 10 Countries with Most Projects

In [32]:
# Find 10 countries with most projects
projects_per_country = world_bank_df.countryname.value_counts()
projects_per_country.head(10)

People's Republic of China         19
Republic of Indonesia              19
Socialist Republic of Vietnam      17
Republic of India                  16
Republic of Yemen                  13
Kingdom of Morocco                 12
People's Republic of Bangladesh    12
Nepal                              12
Republic of Mozambique             11
Africa                             11
Name: countryname, dtype: int64

## Problem 2: Find Top 10 Major Project Themes

In [33]:
# Load JSON file into string and normalize
with open('world_bank_projects.json') as data_file:
    raw_json = json.load(data_file)
project_themes = json_normalize(raw_json, 'mjtheme_namecode', ['id'])

In [34]:
# Group themes by 'id' and list top 10 projects
project_themes.groupby('id').code.nunique().sort_values(ascending=False).head(10)
print('Top 10 Major World Bank Project Themes:')
project_themes.name.value_counts().head(10)

Top 10 Major World Bank Project Themes:


Environment and natural resources management    223
Rural development                               202
Human development                               197
Public sector governance                        184
Social protection and risk management           158
Financial and private sector development        130
                                                122
Social dev/gender/inclusion                     119
Trade and integration                            72
Urban development                                47
Name: name, dtype: int64

## Problem 3: Find and Replace Missing Values

In [35]:
# Create data frame sorted by 'code' and 'name'
project_themes = project_themes.sort_values(['code','name'])
project_themes.head(10)

code                 name       id
212     1                       P128282
363     1                       P144030
1024    1                       P124114
1114    1                       P131094
1437    1                       P130412
2       1  Economic management  P144674
88      1  Economic management  P132425
175     1  Economic management  P118027
204     1  Economic management  P126034
205     1  Economic management  P126034

In [36]:
# Replace missing values with Nan using numpy
project_themes.name[project_themes['name'] == ''] = np.nan
project_themes.head(10)

code                 name       id
212     1                  NaN  P128282
363     1                  NaN  P144030
1024    1                  NaN  P124114
1114    1                  NaN  P131094
1437    1                  NaN  P130412
2       1  Economic management  P144674
88      1  Economic management  P132425
175     1  Economic management  P118027
204     1  Economic management  P126034
205     1  Economic management  P126034

In [37]:
# Backfill all values listed as NaN
project_themes = project_themes.fillna(method='bfill')
project_themes.head(10)

code                 name       id
212     1  Economic management  P128282
363     1  Economic management  P144030
1024    1  Economic management  P124114
1114    1  Economic management  P131094
1437    1  Economic management  P130412
2       1  Economic management  P144674
88      1  Economic management  P132425
175     1  Economic management  P118027
204     1  Economic management  P126034
205     1  Economic management  P126034

## Final Results

| Country | Number of Projects   |
|------|------|
|Republic of Indonesia  | 19|
|People's Republic of China  | 19|
|Socialist Republic of Vietnam  | 17|
|Republic of India  | 16|
|Republic of Yemen  | 13|
|Nepal  | 12|
|Kingdom of Morocco  | 12|
|People's Republic of Bangladesh  | 12|
|Africa  | 12|
|Republic of Mozambique  | 11|

| Project Code | Number of Projects   |
|------|------|
|11  |  250  |
|10  |  216  |
|8   |  210  |
|2   |  199  |
|6   |  168  |
|4   |  146  |
|7   |  130  |
|5   |   77  |
|9   |   50  |
|1   |   38  |